client/src/language/middlewareDefinition.ts

```diff
@@ -0,0 +1,147 @@
+
+import { Location, Position, Range, Uri, commands, type LocationLink, type TextDocument } from 'vscode'
+import { type DefinitionMiddleware } from 'vscode-languageclient'
+
+import { getFileContent } from '../lib/src/utils/files'
+import { requestsManager } from './RequestManager'
+import { changeDefinitionUri, checkIsDefinitionRangeEqual, checkIsDefinitionUriEqual, convertToSameDefinitionType, getDefinitionUri, getEmbeddedLanguageDocPosition, getOriginalDocRange } from './utils'
+import { type EmbeddedLanguageDocInfos } from '../lib/src/types/embedded-languages'
+
+export const middlewareProvideDefinition: DefinitionMiddleware['provideDefinition'] = async (document, position, token, next) => {
+  const nextResult = await next(document, position, token)
+  if ((Array.isArray(nextResult) && nextResult.length !== 0) || (!Array.isArray(nextResult) && nextResult !== undefined)) {
+    return nextResult
+  }
+  const embeddedLanguageDocInfos = await requestsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), position)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return
+  }
+  const embeddedLanguageDocContent = await getFileContent(Uri.parse(embeddedLanguageDocInfos.uri).fsPath)
+  if (embeddedLanguageDocContent === undefined) {
+    return
+  }
+  const adjustedPosition = getEmbeddedLanguageDocPosition(
+    document,
+    embeddedLanguageDocContent,
+    embeddedLanguageDocInfos.characterIndexes,
+    position
+  )
+  const vdocUri = Uri.parse(embeddedLanguageDocInfos.uri)
+  const tempResult = await commands.executeCommand<Location[] | LocationLink[]>(
+    'vscode.executeDefinitionProvider',
+    vdocUri,
+    adjustedPosition
+  )
+
+  if (checkIsArrayLocation(tempResult)) {
+    return await processDefinitions(tempResult, document, embeddedLanguageDocContent, embeddedLanguageDocInfos)
+  } else {
+    return await processDefinitions(tempResult, document, embeddedLanguageDocContent, embeddedLanguageDocInfos)
+  }
+}
+
+const checkIsArrayLocation = (array: Location[] | LocationLink[]): array is Location[] => {
+  return array[0] instanceof Location
+}
+
+const processDefinitions = async <DefinitionType extends Location | LocationLink>(
+  definitions: DefinitionType[],
+  originalTextDocument: TextDocument,
+  embeddedLanguageDocContent: string,
+  embeddedLanguageDocInfos: EmbeddedLanguageDocInfos
+): Promise<DefinitionType[]> => {
+  const result: DefinitionType[] = []
+  await Promise.all(definitions.map(async (definition) => {
+    if (!checkIsDefinitionUriEqual(definition, Uri.parse(embeddedLanguageDocInfos.uri))) {
+      result.push(definition) // only definitions located on the embedded language documents need ajustments
+      return
+    }
+    if (embeddedLanguageDocInfos.language === 'python') {
+      const dDefinition = await getDefinitionOfD(definition)
+      if (dDefinition !== undefined) {
```

Comment:
Could look a bit more tidy if we use a dictionary here? keys would be the symbol of which we are swapping the definition, and value could be the location of the definition? Would look better if we add more of them. (oe?)

Reply:
Makes sense. I did something in the kind.